### PR TITLE
Remove deprecations and other breaking fixes before 3.0 release

### PR DESCRIPTION
**Pull request by vbauzysvmware**
* Fix for https://github.com/vmware/terraform-provider-vcd/issues/355 - resolution was introduce "size_in_mb" which matches what UI allows. In this way it is possible handle API truncation as this field is Int. Other two fields "size" which was existing and "size_in_bytes" which was not working solution and commented out were removed.
* From Edge gtw removed "external_networks, default_gateway_network, advanced"
* From vApp removed "template_name, catalog_name, network_name, memory, cpus, ip, storage_profile, initscript, ovf, accept_all_eulas". Also deprecated field "power_on" which works only in one case - on update and only if vApp already has Vms.
* Removed left over documentation for "vcd_network"
* `virtual_machine_ids` renamed to `vm_ids`

### CODE REVIEW

File: website/docs/d/edgegateway.html.markdown
Comment (Didainius) [inline website/docs/d/edgegateway.html.markdown:32]: Good opportunity to use new notation:
```suggestion
  name = data.vcd_edgegateway.mygw.external_network.name
```

**Review by Didainius — CHANGES_REQUESTED** [review]
Very cool to see things go away :) 

A few general things:
* We need to put user-facing documentation somewhere. Not sure if it is the best place but at least @dataclouder put it here: https://github.com/vmware/terraform-provider-vcd/pull/557/files#diff-12848d8d421c2fa28469f9cfc40602f0R30 cc @lvirbalas , @dataclouder 
* I don't see any skip directives for `upgrade tests`, but this must hit it hard.

**Comment by lvirbalas** [comment]
> We need to put user-facing documentation somewhere. Not sure if it is the best place but at least @dataclouder put it here: https://github.com/vmware/terraform-provider-vcd/pull/557/files#diff-12848d8d421c2fa28469f9cfc40602f0R30 cc @lvirbalas , @dataclouder

@Didainius is the question where to put docs about removed resources? If yes, then the above place in index looks good for that.
CC: @vbauzysvmware

**Comment by Didainius** [comment]
> @Didainius is the question where to put docs about removed resources? If yes, then the above place in index looks good for that.

Yes. Lets put it there then

File: .changes/v3.0.0/558-deprecations.md
Comment (vbauzysvmware) [reply .changes/v3.0.0/558-deprecations.md:1]: fixed

File: vcd/resource_vcd_vm_internal_disk_test.go
Comment (vbauzysvmware) [reply vcd/resource_vcd_vm_internal_disk_test.go:256]: image
I don't why github don't show

File: vcd/resource_vcd_vm_affinity_rule_test.go
Comment (vbauzysvmware) [reply vcd/resource_vcd_vm_affinity_rule_test.go:268]: question for @dataclouder

File: website/docs/d/edgegateway.html.markdown
Comment (vbauzysvmware) [reply website/docs/d/edgegateway.html.markdown]: changed

**Comment by vbauzysvmware** [comment]
> > @Didainius is the question where to put docs about removed resources? If yes, then the above place in index looks good for that.
> 
> Yes. Lets put it there then

I haven't removed any resources, I removed fields. You want me add field list to Index page?

File: .changes/v3.0.0/558-deprecations.md
Comment (lvirbalas) [inline .changes/v3.0.0/558-deprecations.md:2]:
```suggestion
* `resource/vcd_nsxv_firewall_rule` `virtual_machine_ids` deprecated in favor of `vm_ids` [GH-558]
* `resource/vcd_vm_affinity_rule` `virtual_machine_ids` deprecated in favor of `vm_ids` [GH-558]
```

File: vcd/resource_vcd_vapp.go
Comment (lvirbalas) [inline vcd/resource_vcd_vapp.go:131]: Wou wou wou :) Why are we removing `power_on`?
1. It has not been deprecated.
2. This how 2.9 docs read:
> power_on - (Optional) A boolean value stating if this vApp should be powered on. Default is true
3. How can user control power on/off of vApp without this parameter?

File: .changes/v3.0.0/588-removals.md
Comment (lvirbalas) [inline .changes/v3.0.0/588-removals.md:3]: 1. Missing a line for `vcd_network` removal.
2. Please put all this information into docs (same ask as https://github.com/vmware/terraform-provider-vcd/pull/563#discussion_r497367021 ).

File: .changes/v3.0.0/588-removals.md
Comment (lvirbalas) [inline .changes/v3.0.0/588-removals.md:2]: We cannot just remove `size`, which was not deprecated before. We don't want to cause breakages. Could you deprecate `size` the way you did with `virtual_machine_ids`?

**Review by lvirbalas — CHANGES_REQUESTED** [review]
Wow, that's a lot of removed code. Feedback in-line.

File: vcd/datasource_vcd_nsxv_firewall.go
Comment (lvirbalas) [inline vcd/datasource_vcd_nsxv_firewall.go:97]: Shouldn't I get a warning when using this parameter:

```
An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # vcd_nsxv_firewall_rule.demo_nsxv_fw_rule will be updated in-place
  ~ resource "vcd_nsxv_firewall_rule" "demo_nsxv_fw_rule" {
        action          = "accept"
        edge_gateway    = "demo-gw"
        enabled         = true
        id              = "131078"
        logging_enabled = false
        name            = "demo-allow-https"
        rule_tag        = 131078
        rule_type       = "user"

      ~ destination {
            exclude             = false
            gateway_interfaces  = []
            ip_addresses        = [
                "10.150.160.116",
            ]
            ip_sets             = []
            org_networks        = []
          ~ virtual_machine_ids = [
              + "urn:vcloud:vm:b11ec7ce-3953-4bcc-a9a8-25e1ccb4c204",
            ]
            vm_ids              = []
        }

        service {
            port        = "4443"
            protocol    = "tcp"
            source_port = "any"
        }

        source {
            exclude             = false
            gateway_interfaces  = []
            ip_addresses        = [
                "any",
            ]
            ip_sets             = []
            org_networks        = []
            virtual_machine_ids = []
            vm_ids              = []
        }
    }

Plan: 0 to add, 1 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

vcd_nsxv_firewall_rule.demo_nsxv_fw_rule: Modifying... [id=131078]
vcd_nsxv_firewall_rule.demo_nsxv_fw_rule: Still modifying... [id=131078, 10s elapsed]
vcd_nsxv_firewall_rule.demo_nsxv_fw_rule: Modifications complete after 19s [id=131078]

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.
```

File: .changes/v3.0.0/588-removals.md
Comment (vbauzysvmware) [reply .changes/v3.0.0/588-removals.md:3]: vcd_network removed by @dataclouder. I removed leftover docs.

File: vcd/resource_vcd_vapp.go
Comment (vbauzysvmware) [reply vcd/resource_vcd_vapp.go:200]: I reviewed `power_on` and brought back part of it's functionality. Though it works only in one case: you have created vApp and have VMs, then on update you can change value from `power_on=false` (change default value from true->false) to `power_on=true`. This due that vApp resource doesn't have VM's and power_on is meaningless - doesn't do anything. You can see in UI it doesn't allow to change state if vApp in `resolved` state and doesn't have VMs.

File: vcd/resource_vcd_vapp.go
Comment (vbauzysvmware) [reply vcd/resource_vcd_vapp.go:131]: I reviewed `power_on` and brought back part of it's functionality. Though it works only in one case: you have created vApp and have VMs, then on update you can change value from `power_on=false` (change default value from true->false) to `power_on=true`. This due that vApp resource doesn't have VM's and power_on is meaningless - doesn't do anything. You can see in UI it doesn't allow to change state if vApp in `resolved` state and doesn't have VMs.

File: .changes/v3.0.0/588-removals.md
Comment (vbauzysvmware) [reply .changes/v3.0.0/588-removals.md:2]: Well current size doesn't work well - it's a bug -> https://github.com/vmware/terraform-provider-vcd/issues/355. With latest version you ask one - you will get different thing(size). It's possible to leave this field as deprecated, but only very limited implementation - just for `create` - no read(update of size not possible in general). Result will be unexpected result for user with VCD versions we use.

File: vcd/datasource_vcd_nsxv_firewall.go
Comment (vbauzysvmware) [reply vcd/datasource_vcd_nsxv_firewall.go:97]: Seems terraform don't see deprecations in TypeList.

Comment (lvirbalas) [reply vcd/datasource_vcd_nsxv_firewall.go:97]: Let's then print the warning, because otherwise users won't know that they need to switch to vm_ids.

File: .changes/v3.0.0/558-deprecations.md
Comment (vbauzysvmware) [reply .changes/v3.0.0/558-deprecations.md]: changed

File: .changes/v3.0.0/588-removals.md
Comment (vbauzysvmware) [reply .changes/v3.0.0/588-removals.md]: added info to index page

File: vcd/datasource_vcd_nsxv_firewall.go
Comment (vbauzysvmware) [reply vcd/datasource_vcd_nsxv_firewall.go:97]: Add custom warning, only visible on "apply" stage

Comment (lvirbalas) [reply vcd/datasource_vcd_nsxv_firewall.go:97]: After testing, it's clear that there's no clean way to keep the old `virtual_machine_ids` field as deprecated in this case. When you try to clear the value apply ends up cycling on this:
```
  # vcd_nsxv_firewall_rule.demo_nsxv_fw_rule will be updated in-place
  ~ resource "vcd_nsxv_firewall_rule" "demo_nsxv_fw_rule" {
        action          = "accept"
        edge_gateway    = "demo-gw"
        enabled         = true
        id              = "131078"
        logging_enabled = false
        name            = "demo-allow-https"
        rule_tag        = 131078
        rule_type       = "user"
      ~ destination {
            exclude             = false
            gateway_interfaces  = []
            ip_addresses        = [
                "10.150.160.116",
            ]
            ip_sets             = []
            org_networks        = []
          ~ virtual_machine_ids = [
              - "urn:vcloud:vm:b11ec7ce-3953-4bcc-a9a8-25e1ccb4c204",
            ]
            vm_ids              = [
                "urn:vcloud:vm:b11ec7ce-3953-4bcc-a9a8-25e1ccb4c204",
            ]
        }
```

File: .changes/v3.0.0/558-deprecations.md
Comment (lvirbalas) [inline .changes/v3.0.0/558-deprecations.md:2]: These news to appear in that new docs section in index.

File: vcd/datasource_vcd_nsxv_firewall.go
Comment (lvirbalas) [reply vcd/datasource_vcd_nsxv_firewall.go]: Confirmed that the new approach of simply renaming from `virtual_machine_ids` to `vm_ids` works as expected in `vcd_nsxv_firewall_rule`.

**Review by lvirbalas — APPROVED** [review]
LGTM! Just one ask: after merging this and #563 please double check the "Removed resources and fields" doc section that it includes all items as expected.

File: vcd/resource_vcd_vapp.go
Comment (dataclouder) [reply vcd/resource_vcd_vapp.go:131]: I don't like this. This property has a specific functionality: tell the vApp to power on. When we do so, all the VMs will also be powered on. If we remove it, we are taking away an essential functionality. If it doesn't work correctly when the vApp is empty, we can adjust to that in the code.

File: .changes/v3.0.0/588-removals.md
Comment (vbauzysvmware) [reply .changes/v3.0.0/588-removals.md:2]: added separate change log

File: vcd/resource_vcd_vapp.go
Comment (vbauzysvmware) [reply vcd/resource_vcd_vapp.go:131]: Well, as I mentioned currently this functionality works only in one case - on update if you already have VMs. On create it won't work cause you can't have have VMs until vApp is created. Another issue is that VMs have power_on value which will conflict with this value as they do similar things. Also additional issue here https://github.com/vmware/terraform-provider-vcd/issues/537. In summary I haven't changed `power_on` functionality - it is the same as it was, just made it `deprecated` that in future we will work on different approach how to handle actions.

**Review by Didainius — APPROVED** [review]
🤞

**Review by dataclouder — APPROVED** [review]

**Pull request merged**
